proof/refine/ARM/KHeap_R.thy

```diff
@@ -74,65 +74,187 @@ abbreviation (input)
 where
   "set_obj' ptr obj s \<equiv> set_ko' ptr (injectKO obj) s"
 
+(* FIXME: move *)
+(** reader monad **)
+
+(* if state s satisfies P, and if f s returns a value, the value satisfies Q *)
+abbreviation
+  ovalid :: "('s \<Rightarrow> bool) \<Rightarrow> ('s,'a) lookup \<Rightarrow> ('a \<Rightarrow> 's \<Rightarrow>  bool) \<Rightarrow> bool"
```

Comment:
There is already a definition of `ovalid` in `lib/Monad_WP/OptionMonadWP.thy`. We should use that one if we need it.

I'd probably state lemmas about the option monad directly in the `[| P s; f s = Some v |] ==> Q s` form, because it's almost zero notational overhead over `ovalid` and more directly applicable as rule. Although we should probably think about this a bit more carefully before we start it everywhere, because we do lose the syntactic hooks that `ovalid` would provide.  (We did use it the "lightweight" way in RISCV).

Reply:
Using the existing `ovalid` instead of what is defined here works fine.

I am wondering if the second part is something we need to discuss further, or something we should try now. I can push an updated version after integrating the `OptionMonadWP` definition, and we can see what it looks like, maybe.

Reply:
It'd be fine to do that later. Maybe make a Jira issue so we don't forget about it.

Reply:
Ok. I guess the point is how to make the definitions nicely work with what `wp` will leave us with. The definition of `ovalid` in `OptionMonadWP` or here make sense conceptually, but they would always need certain steps before being able to apply to what you get after applying `wp`.

My solution for now is to use this one
```
lemma use_ovalid:
  "[| ovalid P f Q; f s = Some r; P s |] ==> Q r s”
```
which seems to work fine.

Reply:
Yes, that looks good.